src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs

```diff
@@ -0,0 +1,185 @@
+using ILLink.Shared.DataFlow;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.FlowAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace ILLink.RoslynAnalyzer.DataFlow
+{
+	// Visitor which tracks the values of locals in a block. It provides extension points that get called
+	// whenever a value that comes from a tracked local reference flows into one of the following:
+	// - field
+	// - parameter
+	// - method return
+	public abstract class LocalDataFlowVisitor<TValue, TValueLattice> : OperationVisitor<LocalState<TValue>, TValue>,
+		ITransfer<BlockProxy, LocalState<TValue>, LocalStateLattice<TValue, TValueLattice>>
+		where TValue : IEquatable<TValue>
+		where TValueLattice : ILattice<TValue>
+	{
+		protected readonly LocalStateLattice<TValue, TValueLattice> LocalStateLattice;
+
+		protected readonly OperationBlockAnalysisContext Context;
+
+		protected TValue TopValue => LocalStateLattice.Lattice.ValueLattice.Top;
+
+		public LocalDataFlowVisitor (LocalStateLattice<TValue, TValueLattice> lattice, OperationBlockAnalysisContext context) =>
+			(LocalStateLattice, Context) = (lattice, context);
+
+		public void Transfer (BlockProxy block, LocalState<TValue> state)
+		{
+			foreach (IOperation operation in block.Block.Operations)
+				Visit (operation, state);
+
+			// Blocks may end with a BranchValue computation. Visit the BranchValue operation after all others.
+			IOperation? branchValueOperation = block.Block.BranchValue;
+			if (branchValueOperation != null) {
+				var branchValue = Visit (branchValueOperation, state);
+
+				// BranchValue may represent a value used in a conditional branch to the ConditionalSuccessor - if so, we are done.
+
+				// If not, the BranchValue represents a return or throw value associated with the FallThroughSuccessor of this block.
+				// (ConditionalSuccessor == null iff ConditionKind == None).
+				if (block.Block.ConditionKind == ControlFlowConditionKind.None) {
+					// This means it's a return value or throw value associated with the fall-through successor.
+
+					// Return statements with return values are represented in the control flow graph as
+					// a branch value operation that computes the return value. The return operation itself is the parent
+					// of the branch value operation.
+
+					// Get the actual "return Foo;" operation (not the branch value operation "Foo").
+					// This should be used as the location of the warning. This is important to provide the right
+					// warning location and because warnings are disambiguated based on the operation.
+					var parentSyntax = branchValueOperation.Syntax.Parent;
+					if (parentSyntax == null)
+						throw new InvalidOperationException ();
+
+					var parentOperation = Context.Compilation.GetSemanticModel (branchValueOperation.Syntax.SyntaxTree).GetOperation (parentSyntax);
+
+					if (parentOperation is IThrowOperation)
+						throw new NotImplementedException ();
```

Comment:
Should we track a TODO for this somewhere?
I don't think we have tests like this in the linker tests.

Reply:
Or maybe track the TODO by adding a linker test for this - eventually we will pick it up in analyzer and find this hole :-)

Reply:
Added a comment